rust/datafusion/src/execution/table_impl.rs

```diff
@@ -78,24 +108,210 @@ impl Table for TableImpl {
+            Some((i, _)) => Ok(i),
+            _ => Err(ExecutionError::InvalidColumn(format!(
+                "No column named '{}'",
+                name
+            ))),
+        }
+    }
+
+    /// Create an expression to represent the max() aggregate function
+    fn min(&self, expr: &Expr) -> Result<Expr> {
+        self.aggregate_expr("MIN", expr)
+    }
+
+    /// Create an expression to represent the max() aggregate function
+    fn max(&self, expr: &Expr) -> Result<Expr> {
+        self.aggregate_expr("MAX", expr)
+    }
+
+    /// Create an expression to represent the max() aggregate function
+    fn sum(&self, expr: &Expr) -> Result<Expr> {
+        self.aggregate_expr("SUM", expr)
+    }
+
+    /// Create an expression to represent the max() aggregate function
+    fn avg(&self, expr: &Expr) -> Result<Expr> {
+        self.aggregate_expr("AVG", expr)
+    }
+
+    /// Create an expression to represent the max() aggregate function
+    fn count(&self, expr: &Expr) -> Result<Expr> {
+        self.aggregate_expr("COUNT", expr)
+    }
+
     /// Convert to logical plan
     fn to_logical_plan(&self) -> Arc<LogicalPlan> {
         self.plan.clone()
     }
 }
 
-/// Create a new schema by applying a projection to this schema's fields
-fn projection(schema: &Schema, projection: &Vec<usize>) -> Result<Arc<Schema>> {
-    let mut fields: Vec<Field> = Vec::with_capacity(projection.len());
-    for i in projection {
-        if *i < schema.fields().len() {
-            fields.push(schema.field(*i).clone());
-        } else {
-            return Err(ExecutionError::InvalidColumn(format!(
-                "Invalid column index {} in projection",
-                i
-            )));
+impl TableImpl {
+    /// Determine the data type for a given expression
+    fn get_data_type(&self, expr: &Expr) -> Result<DataType> {
+        match expr {
+            Expr::Column(i) => Ok(self.plan.schema().field(*i).data_type().clone()),
+            _ => Err(ExecutionError::General(format!(
+                "Could not determine data type for expr {:?}",
+                expr
+            ))),
         }
     }
-    Ok(Arc::new(Schema::new(fields)))
+
+    /// Create an expression to represent a named aggregate function
+    fn aggregate_expr(&self, name: &str, expr: &Expr) -> Result<Expr> {
+        let return_type = self.get_data_type(expr)?;
+        Ok(Expr::AggregateFunction {
+            name: name.to_string(),
+            args: vec![expr.clone()],
+            return_type,
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::execution::context::ExecutionContext;
+    use std::env;
+
+    #[test]
+    fn column_index() {
+        let t = test_table();
+        assert_eq!(0, t.column_index("c1").unwrap());
+        assert_eq!(1, t.column_index("c2").unwrap());
+        assert_eq!(12, t.column_index("c13").unwrap());
+    }
+
+    #[test]
+    fn select_columns() -> Result<()> {
```

Comment:
Nice to know we can return `Result<()>` here. We can avoid lots of `unwrap`s in all the unit tests we have.